### PR TITLE
Reduce the maximum size of input splits in Flink to better distribute work

### PR DESCRIPTION
**Pull request by jto**
In the current implementation of file sources, the Flink runner will try to fix the size of input splits using the formula `source estimated size / parallelism`. Most of the time, the resulting desired split size will be much larger than the size of files read by this source.  In that case each individual file becomes full split, and the number of splits is larger than parallelism. 

This creates an issue when the size of individual files varies a lot. At Spotify we commonly read datasets stored on GCS where individual files range from a few Mb to 1 or 2 Gb. Because of this, work distribution is sub-optimal because some workers will be consuming larges splits while other worker consume smaller splits.  

This PR limits the maximum size of each split to 128Mb (configurable), which is a decent tradeoff between balancing work and limiting the overhead of consuming splits. I believe Dataflow uses the same strategy.

In our test on real workflows, we measured an **improvement in execution time (and therefore resource consumption) of up to 20%**. Of course the overall improvement varies a lot from one workflow to another, but we never encountered a case where performances where noticeably degraded.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by robertwb** [comment]
If I understand correctly, you're setting this cap to reduce the ratio between the largest and smallest shard? This seems a reasonable workaround to handle pathologically bad situations. 

(BTW, the way that Dataflow handles this is that it is not limited by the initial splitting, but rather creates further splits dynamically when some workers have finished their chunk and others have substantial work left to do. This is especially handy when sources are not byte based (or the processing is not evenly spread out among the bytes).)

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/SourceInputFormat.java [hunk 1]
```diff
@@ -108,12 +110,27 @@ public float getAverageRecordWidth() {
     return null;
   }
 
+  private long getDesiredSizeBytes(int numSplits) throws Exception {
+    long totalSize = initialSource.getEstimatedSizeBytes(options);
+    long defaultSplitSize = totalSize / numSplits;
+    if (initialSource instanceof FileBasedSource) {
+      long maxSplitSize = options.as(FlinkPipelineOptions.class).getFileInputSplitMaxSizeBytes();
```
Comment (robertwb) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/SourceInputFormat.java:117]: Could we let a value of 0 or -1 disable this new behavior and restore to the original?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 2]
```diff
@@ -313,6 +313,12 @@ public interface FlinkPipelineOptions
 
   void setFlinkConfDir(String confDir);
 
+  @Description("Set the maximum size of input split when data is read from a filesystem.")
+  @Default.Long(128 * 1024 * 1024)
```
Comment (robertwb) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java:317]: We could let the default be the old behavior, and only set this for pipelines that are affected. If you have good evidence that this is generally helpful and rarely hurtful providing a good default makes sense.

**Review by robertwb — APPROVED** [review]
I do think we should make a note about this in the top-level CHANGES.md file.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/SourceInputFormat.java (hunk 1 above)
Comment (jto) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/SourceInputFormat.java:117]: Sounds reasonable. I'll make the change.

**Comment by jto** [comment]
Thanks for the review @robertwb.  I made the suggested changes and also changed the option to be in Mb directly for easier config.

**Comment by robertwb** [comment]
Thanks. Looks good.

**Pull request merged**